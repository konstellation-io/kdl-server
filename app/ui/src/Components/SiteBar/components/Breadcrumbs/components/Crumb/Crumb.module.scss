@import 'Styles/variables';
@import 'Styles/mixins';
@import 'Styles/colors';

.container {
  position: relative;
  display: flex;
  flex-direction: row;
  align-items: center;
  height: 100%;
  cursor: pointer;

  transition: background-color ease 0.3s;

  &:not(&.select):hover {
    background-color: palette(base, 400);
  }


  &.hidden {
    display: none;
  }

  .crumbTitle {
    color: font-color(light);
    font-size: small;
    margin-right: 5px;
  }

  .crumbContainer {
    display: flex;
    align-items: center;
    height: 100%;
    width: 100%;

    &.select {
      border-radius: $grid-unit;
      background-color: palette(base);
      padding: $grid-unit;
      height: 80%;

      .crumbText {
        @include font-body;
        color: font-color(grey);
        font-weight: bold;
        margin-left: $grid-unit;
      }
    }

    .crumbText {
      @include font-body;
      color: font-color(light);
      margin-left: $grid-unit;
    }
    .crumbBar {
      margin-left: 2 * $grid-unit;
    }
  }
  .content {
    display: flex;
    flex-direction: column;
    position: absolute;
    top: 5 * $grid-unit;
    left: 0;
    z-index: 3;
    background-color: palette(base);
    cursor: initial;
    min-width: 100%;
  }
  .rightIcon {
    margin-left: 2 * $grid-unit;
    transition: transform ease 0.3s;
    &.opened {
      transform: scaleY(-1);
    }
  }
}
