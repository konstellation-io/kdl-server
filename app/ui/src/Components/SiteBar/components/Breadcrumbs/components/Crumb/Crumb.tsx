import * as React from 'react';
import styles from './Crumb.module.scss';
import cx from 'classnames';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
import useBoolState from 'Hooks/useBoolState';
import { ExpandableMenu } from 'kwc';

export interface BottomComponentProps {
  closeComponent: () => void;
}
export type CrumbProps = {
  dataTestId?: string;
  crumbText: string;
  LeftIconComponent: React.ReactElement;
  children: (props: BottomComponentProps) => JSX.Element;
};

function Crumb({ crumbText, LeftIconComponent, dataTestId, children }: CrumbProps) {
  const { value: opened, toggle: toggleComponent, deactivate: hideComponent } = useBoolState(false);

  return (
    <div className={styles.container} onClick={toggleComponent} data-testid={dataTestId}>
      <div className={styles.crumbContainer}>
        {LeftIconComponent}
        <span className={styles.crumbText}>{crumbText}</span>
        <ExpandMoreIcon
          className={cx(styles.rightIcon, 'icon-regular', {
            [styles.opened]: opened,
          })}
        />
      </div>
      <ExpandableMenu opened={opened} close={hideComponent} className={styles.content}>
        {children({
          closeComponent: hideComponent,
        })}
      </ExpandableMenu>
    </div>
  );
}

export default Crumb;
