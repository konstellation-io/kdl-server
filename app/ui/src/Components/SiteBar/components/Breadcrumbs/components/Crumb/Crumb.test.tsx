import Crumb, { CrumbProps } from './Crumb';
import React from 'react';
import { shallow } from 'enzyme';
import IconHome from '@material-ui/icons/Dashboard';
import AnimateHeight from 'react-animate-height';

const BottomComponent = () => <div>bar</div>;
const crumbProps: CrumbProps = {
  crumbText: 'foo',
  BottomComponent,
  LeftIconComponent: <IconHome />,
};

let component;
beforeEach(() => {
  component = shallow(<Crumb {...crumbProps} />);
});

describe('Crumb component', () => {
  it('should render without crashing', () => {
    expect(component).toMatchSnapshot();
  });

  describe('behavior', () => {
    it('should show the bottom component when click on crumb', () => {
      // Arrange.
      const container = component.find('.container');
      const contentProps = component.find(AnimateHeight).props();
      const heightBeforeClick = contentProps.height;

      // Act.
      container.simulate('click');

      // Assert.
      const contentPropsAfter = component.find(AnimateHeight).props();
      const heightAfterClick = contentPropsAfter.height;
      expect(heightBeforeClick).toBe(0);
      expect(heightAfterClick).toBe('auto');
    });

    it('should hide the bottom component when double click on crumb', () => {
      // Arrange.
      const heightBeforeClick = component.find(AnimateHeight).props().height;

      // Act.
      component.find('.container').simulate('click');
      component.find('.container').simulate('click');

      // Assert.
      const heightAfterClick = component.find(AnimateHeight).props().height;
      expect(heightBeforeClick).toBe(0);
      expect(heightAfterClick).toBe(0);
    });
  });
});
