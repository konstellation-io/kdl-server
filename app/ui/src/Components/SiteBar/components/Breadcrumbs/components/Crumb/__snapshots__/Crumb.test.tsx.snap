// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Crumb component should render without crashing 1`] = `
<div
  className="container"
>
  <div
    className="crumbContainer"
    onClick={[Function]}
  >
    <Memo(ForwardRef(DashboardIcon)) />
    <span
      className="crumbText"
    >
      foo
    </span>
    <Memo(ForwardRef(ExpandMoreIcon))
      className="rightIcon icon-regular"
    />
  </div>
  <N
    className="content"
    close={[Function]}
    opened={false}
  >
    <div>
      bar
    </div>
  </N>
</div>
`;
